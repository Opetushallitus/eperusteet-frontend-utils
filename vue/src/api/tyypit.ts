import * as EperusteetApi from '../generated/eperusteet/api';
import * as YlopsApi from '../generated/ylops/api';


export interface LokalisoituTekstiDto {
  // id?: number;
  // tunniste?: string;
  [key: string]: string;
}

export interface ViiteLaaja extends EperusteetApi.PerusteenOsaViiteDto {
  lapset?: Array<object>;
}


export type PerusteDto = EperusteetApi.PerusteDto;
export type PerusteHakuDto = EperusteetApi.PerusteHakuDto;
export type TiedoteDto = EperusteetApi.TiedoteDto;
export type PageTiedoteDto = EperusteetApi.PageTiedoteDto;
export type PerusteKoosteDto = EperusteetApi.PerusteKoosteDto;
export type Lops2019SisaltoDto = EperusteetApi.Lops2019SisaltoDto;
export type PerusteenOsaViiteDtoObject = EperusteetApi.PerusteenOsaViiteDtoObject;
export type Matala = EperusteetApi.Matala;
export type Laaja = EperusteetApi.Laaja;
export type Lops2019OppiaineDto = EperusteetApi.Lops2019OppiaineDto;
export type Lops2019OppiaineKaikkiDto = EperusteetApi.Lops2019OppiaineKaikkiDto;
export type Lops2019ModuuliDto = EperusteetApi.Lops2019ModuuliDto;
export type Lops2019LaajaAlainenOsaaminenKokonaisuusDto = EperusteetApi.Lops2019LaajaAlainenOsaaminenKokonaisuusDto;
export type NavigationNodeDto = EperusteetApi.NavigationNodeDto;

export type OpetussuunnitelmaDto = YlopsApi.OpetussuunnitelmaDto
export type OpetussuunnitelmaInfoDto = YlopsApi.OpetussuunnitelmaInfoDto;
export type OpetussuunnitelmaJulkinenDto = YlopsApi.OpetussuunnitelmaJulkinenDto;
export type OpetussuunnitelmaKevytDto = YlopsApi.OpetussuunnitelmaKevytDto;
