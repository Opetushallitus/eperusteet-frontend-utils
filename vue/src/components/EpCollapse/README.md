EpCollapse:

```vue

<ep-collapse>
    <div slot="header">
        Hi, i'm a header!
    </div>
    <div>
        Hi, i'm the content!
    </div>
</ep-collapse>

```
