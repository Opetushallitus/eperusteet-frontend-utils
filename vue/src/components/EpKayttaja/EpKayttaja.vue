<template>
<div class="kayttaja">
  <b-nav-item-dropdown id="kayttaja-dropdown" right no-caret>
    <template slot="button-content">
      <div class="d-flex flex-row">
        <div class="kayttaja-valikko d-flex flex-column">
          <span class="kielivalitsin text-right">{{ esitysnimi }}</span>
          <small v-if="koulutustoimija" class="valittu-koulutustoimija text-right">{{ $kaanna(koulutustoimija.nimi) }}</small>
        </div>
        <EpMaterialIcon>expand_more</EpMaterialIcon>
      </div>
    </template>

    <!-- Koulutustoimija -->
    <ep-collapse :expanded-by-default="false" v-if="koulutustoimija" :use-padding="false" :border-bottom="false">
      <div slot="header">
        <div class="pl-3 pt-2 text-nowrap kieli">
          <EpMaterialIcon class="icon mr-3">group</EpMaterialIcon>
          <span>{{ $t('organisaatio') }}</span>
        </div>
        <div class="pl-3 text-nowrap">
          <span class="icon mr-3" />
          <small>{{ $kaanna(koulutustoimija.nimi) }}</small>
        </div>
      </div>

      <template slot="icon" slot-scope="{ toggled }">
        <div class="ml-auto align-self-start" style="padding: 0.8rem 1rem;">
          <EpMaterialIcon v-if="toggled">expand_less</EpMaterialIcon>
          <EpMaterialIcon v-else>expand_more</EpMaterialIcon>
        </div>
      </template>

      <div class="collapse-tausta text-left mt-2">
        <template v-if="koulutustoimijat && koulutustoimijat.length > 10">
          <hr class="m-0"/>
          <EpSearch class="rajain pl-2" v-model="koulutustoimijaQuery" />
          <hr class="mt-0"/>
        </template>

        <div class="koulutustoimijat mb-2">
          <div class="d-flex justify-content-end" v-if="hasLukuOikeusKoulutustoimijoita">
            <EpMaterialIcon class="vain-luku mr-3">visibility</EpMaterialIcon>
            <ep-toggle v-model="naytaLukuoikeusKoulutustoimijat" :title="$t('lukuoikeus')"/>
          </div>
          <b-dd-item-button @click="valitseOrganisaatio(kt)"
                            v-for="kt in koulutustoimijatFilteredSorted"
                            :key="kt.id"
                            :disabled="koulutustoimija.id === kt.id">
            <div class="row">
              <div class="collapse-tausta-valinta-icon col-1">
                <EpMaterialIcon v-if="koulutustoimija === kt.id" class="valittu">done</EpMaterialIcon>
              </div>
              <div class="col-10 koulutustoimija">
                {{ $kaanna(kt.nimi) }}
                <EpMaterialIcon v-if="kt.oikeus === 'luku'" class="vain-luku" size="16px">visibility</EpMaterialIcon>
              </div>
            </div>
          </b-dd-item-button>
        </div>
      </div>
    </ep-collapse>

    <!-- Käyttöliittymän kieli -->
    <ep-collapse :expanded-by-default="false" :use-padding="false" :border-bottom="false">
      <div slot="header">
        <div class="pl-3 pt-2 text-nowrap kieli">
          <EpMaterialIcon class="icon mr-3">language</EpMaterialIcon>
          <span>{{ $t('kieli') }}</span>
        </div>
        <div class="pl-3 uikieli">
          <span class="icon mr-3" />
          <small>{{ $t(uiKieli) }}</small>
        </div>
      </div>

      <template slot="icon" slot-scope="{ toggled }">
        <div class="ml-auto align-self-start" style="padding: 0.8rem 1rem;">
          <EpMaterialIcon v-if="toggled">expand_less</EpMaterialIcon>
          <EpMaterialIcon v-else>expand_more</EpMaterialIcon>
        </div>
      </template>

      <div class="collapse-tausta text-left">
        <b-dd-item-button @click="valitseUiKieli(kieli)"
                          v-for="kieli in sovelluksenKielet"
                          :key="kieli"
                          :disabled="kieli === uiKieli">
          <div class="collapse-tausta-valinta-icon">
            <EpMaterialIcon v-if="kieli === uiKieli" class="mr-3 valittu">done</EpMaterialIcon>
          </div>
          {{ $t(kieli) }}
        </b-dd-item-button>
      </div>
    </ep-collapse>

    <b-dd-item href="/henkilo-ui/omattiedot">
      <EpMaterialIcon class="icon mr-3">person</EpMaterialIcon>
      <span>{{ $t('kayttajan-asetukset') }}</span>
    </b-dd-item>

    <b-dropdown-divider/>

    <b-dd-item href="/virkailijan-tyopoyta" v-if="!sovellusOikeudet || sovellusOikeudet.length === 1">
      <EpMaterialIcon class="icon mr-3">launch</EpMaterialIcon>
      <span>{{ $t('palaa-virkailijan-tyopyodalle') }}</span>
    </b-dd-item>

    <!-- Sovellussiisrtymä  -->
    <ep-collapse :expanded-by-default="false" :use-padding="false" :border-bottom="false" v-else>
      <div slot="header">
        <div class="pl-3 pt-2 text-nowrap kieli">
          <EpMaterialIcon class="icon mr-3">launch</EpMaterialIcon>
          <span>{{ $t('vaihda-sovellusta') }}</span>
        </div>
        <div class="pl-3 valittu-sovellus pb-2" v-if="valittuSovellus">
          <span class="icon mr-3" />
          <small>{{ $t(valittuSovellus.eperusteSovellus.sovellus) }}</small>
        </div>
      </div>

      <template slot="icon" slot-scope="{ toggled }">
        <div class="ml-auto align-self-start" style="padding: 0.8rem 1rem;">
          <EpMaterialIcon v-if="toggled">expand_less</EpMaterialIcon>
          <EpMaterialIcon v-else>expand_more</EpMaterialIcon>
        </div>
      </template>

      <div class="collapse-tausta text-left">
        <b-dd-item :href="sovellusOikeus.eperusteSovellus.url"
                    v-for="sovellusOikeus in sovellusOikeudet"
                    :key="sovellusOikeus.eperusteSovellus.sovellus"
                    :disabled="sovellusOikeus.valittu"
                    class="sovellusoikeus">
          <div class="collapse-tausta-valinta-icon">
            <EpMaterialIcon v-if="sovellusOikeus.valittu" class="mr-3 valittu">done</EpMaterialIcon>
          </div>
          {{ $t(sovellusOikeus.eperusteSovellus.sovellus) }}
        </b-dd-item>

        <b-dd-item href="/virkailijan-tyopoyta">
          <div class="collapse-tausta-valinta-icon" />
          {{ $t('virkailijan-tyopoyta') }}
        </b-dd-item>
      </div>
    </ep-collapse>

    <b-dropdown-divider/>

    <b-dd-item :href="logoutHref">
      <EpMaterialIcon class="mr-3 valittu">logout</EpMaterialIcon>
      <span>{{ $t('kirjaudu-ulos') }}</span>
    </b-dd-item>

  </b-nav-item-dropdown>
</div>
</template>

<script lang="ts">
import * as _ from 'lodash';
import { Prop, Component, Vue } from 'vue-property-decorator';
import { Kielet, UiKielet } from '../../stores/kieli';
import { Kieli } from '../../tyypit';
import { IEsitysnimi, parsiEsitysnimi } from '../../utils/kayttaja';
import EpCollapse from '../EpCollapse/EpCollapse.vue';
import EpSpinner from '../EpSpinner/EpSpinner.vue';
import { setItem } from '@shared/utils/localstorage';
import { SovellusOikeus } from '@shared/plugins/oikeustarkastelu';
import EpSearch from '@shared/components/forms/EpSearch.vue';
import EpToggle from '@shared/components/forms/EpToggle.vue';
import EpMaterialIcon from '@shared/components/EpMaterialIcon/EpMaterialIcon.vue';

@Component({
  components: {
    EpCollapse,
    EpSpinner,
    EpSearch,
    EpToggle,
    EpMaterialIcon,
  },
})
export default class EpKayttaja extends Vue {
  @Prop({ required: true })
  private tiedot!: IEsitysnimi;

  @Prop({})
  private koulutustoimija!: any | null;

  @Prop({})
  private koulutustoimijat!: any[] | null;

  @Prop({ required: false })
  private sovellusOikeudet!: SovellusOikeus[];

  @Prop({ required: false, default: '/service-provider-app/saml/logout' })
  private logoutHref!: string;

  private koulutustoimijaQuery = '';
  private naytaLukuoikeusKoulutustoimijat = true;

  get esitysnimi() {
    return parsiEsitysnimi(this.tiedot);
  }

  get sovelluksenKielet() {
    return UiKielet;
  }

  get uiKieli() {
    return Kielet.uiKieli.value;
  }

  get hasLukuOikeusKoulutustoimijoita() {
    if (this.koulutustoimijat) {
      return !_.isEmpty(_.filter(this.koulutustoimijat, { oikeus: 'luku' })) && !_.isEmpty(_.reject(this.koulutustoimijat, { oikeus: 'luku' }));
    }
  }

  get koulutustoimijatFilteredSorted() {
    return _.chain(this.koulutustoimijat)
      .filter(kt => Kielet.search(this.koulutustoimijaQuery, kt.nimi))
      .filter(kt => this.naytaLukuoikeusKoulutustoimijat || kt.oikeus !== 'luku')
      .map(kt => {
        return {
          ...kt,
          kaannettyNimi: this.$kaanna(kt.nimi),
        };
      })
      .orderBy(['kaannettyNimi', 'id'], ['asc', 'asc'])
      .value();
  }

  private async valitseOrganisaatio(koulutustoimija: any) {
    if (!this.$router) {
      return;
    }

    const router = this.$router;
    const current: any = router.currentRoute;
    const next = {
      ...current,
      params: {
        ...current.params,
        koulutustoimijaId: _.toString(koulutustoimija.id),
      },
    };

    try {
      await router.push(next);
      setItem('koulutustoimija', koulutustoimija.id);
    }
    catch (err) { }
  }

  private async valitseUiKieli(kieli: Kieli) {
    Kielet.setUiKieli(kieli);

    if (!this.$router) {
      return;
    }

    const router = this.$router;
    const current: any = router.currentRoute;
    const next = {
      ...current,
      params: {
        ...current.params,
        lang: kieli || this.$i18n.fallbackLocale,
      },
    };

    try {
      await router.push(next);
    }
    catch (err) { }
  }

  get valittuSovellus() {
    return _.find(this.sovellusOikeudet, 'valittu');
  }
}
</script>

<style scoped lang="scss">
@import '~@shared/styles/_variables.scss';
.kayttaja {

  ::v-deep ul.dropdown-menu {
    width: 350px;
  }

  .icon {
    display: inline-block;
    width: 1rem;
  }

  .kayttaja-valikko {
    color: white;
  }

  .valittu-koulutustoimija {
    font-size: 0.75rem;
  }

  .rajain {
    background-color: $white;

    ::v-deep .form-control {
      background-color: $white;
    }
  }

  .koulutustoimijat {
    overflow-y: auto;
    max-height: 400px;

    .vain-luku {
      color: $blue;
    }

    .koulutustoimija {
      white-space:normal;
      line-height: 1.2rem;
    }
  }

  /deep/ .ep-collapse {
    padding-top: 0;
    padding-bottom: 0;
  }

  .collapse-tausta {
    background-color: #F3F3F3;
    text-align: right;
    white-space: nowrap;
    overflow: hidden;

    /deep/ .dropdown-item {
      padding: 0.25rem 1rem;
    }

    .valittu {
      color: #3467E3;
      vertical-align: -0.25em;
    }

    .collapse-tausta-valinta-icon {
      display: inline-block;
      width: 28px;
    }
  }

  .kieli {
    font-weight: normal;
  }

  /deep/ .dropdown-menu {
    padding: 0;
    color: #000000;
  }

  /deep/ .dropdown-divider {
    margin: 0;
  }

  /deep/ .dropdown-item {
    padding: 0.5rem 1rem;
    color: #000000;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  /deep/ .dropdown-item:disabled {
    color: inherit;
  }

  /deep/ .dropdown-item:hover {
    background-color: inherit;
  }

  /deep/ .dropdown-item:active {
    color: inherit;
    background-color: inherit;
  }
}
</style>
