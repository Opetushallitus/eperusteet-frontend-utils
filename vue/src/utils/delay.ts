import * as _ from 'lodash';
import Vue from 'vue';
import VueCompositionApi, { computed } from '@vue/composition-api';

Vue.use(VueCompositionApi);

/**
 * Asyncronic debounce that supports awaiting. Useful when need to wait
 * for the actual event to happen.
 *
 * @param {async function} fn Function to be called after debounce
 * @param {number} ms Time to wait in milliseconds
 * @returns {Promise} Is resolved after the actual debounce function runs
 */
export function asyncDebounce<T, F>(fn: F & any, ms: number): F & any {
  let tid = null as number | null;
  const wrapper = function(this: any, ...params: any[]) {
    if (_.isNumber(tid)) {
      clearTimeout(tid);
    }
    const self = this;
    return new Promise((resolve, reject) => {
      tid = setTimeout(async () => {
        tid = null;
        try {
          const bound = fn.bind(self);
          resolve(await bound(params));
        }
        catch (err) {
          reject(err);
        }
      }, ms);
    });
  };
  return wrapper as unknown as F;
}

/**
 * Debounced that can be used for functions as a decorator.
 *
 * + Can be awaited
 * + Can be used in multiple instances of the same class
 *
 * @param {number} ms
 */
export function Debounced(ms = 300) {
  const debounces = new WeakMap();
  return function(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const original = descriptor.value;
    descriptor.value = async function(this: any, ...params: any[]) {
      if (debounces.has(this)) {
        clearTimeout(debounces.get(this));
      }
      return new Promise((resolve, reject) => {
        if (process.env.NODE_ENV === 'test') {
          original.apply(this, params).then(resolve)
            .catch(reject);
        }
        else {
          debounces.set(this, setTimeout(() => {
            original.apply(this, params).then(resolve)
              .catch(reject);
            debounces.set(this, undefined);
          }, ms));
        }
      });
    };
  };
}

/**
 * Wraps a getter into a computed value
 */
export function Computed() {
  return function(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    if (!_.isFunction(descriptor.get)) {
      throw new Error('descriptor must have a getter');
    }
    const original = descriptor.get;
    const getter = computed(() => original);
    descriptor.value = {
      get: getter.value,
    };
  };
}

/**
 * delay
 *
 * @param ms
 */
export async function delay(ms: number = 0) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}
