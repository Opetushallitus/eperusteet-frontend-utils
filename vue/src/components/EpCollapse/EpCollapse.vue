<template>
  <div>
    <hr v-if="first && !borderTop" />
    <div :class="classess" v-if="!disableHeader" :style="styles.collapse" @click="togglefull ? toggle(): null" @keyup.enter="togglefull ? toggle(): null">
      <!-- Button tagia ei voida käyttää, sillä ml-auto ei toimi.-->
      <!-- Käytetään button rolea saavutettavuuden takaamiseksi.-->
      <div v-if="hasHeaderSlot"
           class="collapse-button d-flex align-items-center mb-2"
           @click="!togglefull ? toggle(): null"
           @keyup.enter="!togglefull ? toggle(): null"
           role="button"
           tabindex="0"
           :aria-expanded="toggled"
           :style="styles.header">
        <slot name="icon" :toggled="toggled" v-if="chevronLocation === 'left' && collapsable">
          <div class="align-self-start mr-2">
            <EpMaterialIcon v-if="toggled" size="28px">expand_less</EpMaterialIcon>
            <EpMaterialIcon v-else size="28px">expand_more</EpMaterialIcon>
          </div>
        </slot>
        <div class="align-self-start header">
          <div :class="{'header-toggled': toggled}">
            <slot name="header" :toggled="toggled"></slot>
          </div>
        </div>
        <slot name="icon" :toggled="toggled" v-if="chevronLocation === 'right'  && collapsable">
          <div class="ml-auto">
            <EpMaterialIcon v-if="toggled" size="28px">expand_less</EpMaterialIcon>
            <EpMaterialIcon v-else size="28px">expand_more</EpMaterialIcon>
          </div>
        </slot>
      </div>
      <div v-if="toggled">
        <slot></slot>
      </div>
    </div>
    <slot v-else></slot>
    <hr v-if="borderBottom" />
  </div>
</template>

<script lang="ts">
import _ from 'lodash';
import { Vue, Component, Prop } from 'vue-property-decorator';
import { setItem, getItem } from '../../utils/localstorage';
import EpMaterialIcon from '@shared/components/EpMaterialIcon/EpMaterialIcon.vue';

@Component({
  components: {
    EpMaterialIcon,
  },
})
export default class EpCollapse extends Vue {
  @Prop({ default: true })
  private expandedByDefault!: boolean;

  @Prop({ default: false })
  private disableHeader!: boolean;

  @Prop({ default: true })
  private usePadding!: boolean;

  @Prop({ default: '' })
  private tyyppi!: string;

  @Prop({ default: false })
  private borderTop!: boolean;

  @Prop({ default: true })
  private borderBottom!: boolean;

  @Prop({ default: 'right' })
  private chevronLocation!: 'right' | 'left';

  @Prop({ default: true })
  private collapsable!: boolean;

  @Prop({ default: false })
  private first!: boolean;

  @Prop({ default: false, type: Boolean })
  private blue!: Boolean;

  @Prop({ required: false, default: false })
  private shadow!: boolean;

  @Prop({ required: false, default: false })
  private togglefull!: boolean;

  private toggled = false;

  get hasHeaderSlot() {
    return !!this.$scopedSlots.header;
  }

  get styles() {
    let style = {
      header: {},
      collapse: {},
    };

    if (this.blue) {
      style = {
        header: {
          'color': '#001A58',
        },
        collapse: {
          'padding': '20px 20px 0px 20px',
          'border-radius': '30px',
          'border': '1px solid #C8F1FF',
          'background': '#E6F6FF',
        },
      };
    }

    if (this.usePadding) {
      style = {
        header: {
          ...style.header,
          'margin-bottom': '10px',
          'margin-top': '10px',
        },
        collapse: {
          ...style.collapse,
          'padding-top': '20px',
          'padding-bottom': '20px',
        },
      };
    }

    return style;
  }

  get classess() {
    let result = 'ep-collapse';
    if (this.borderTop) {
      result += ' topborder';
    }

    if (this.shadow) {
      result += ' shadow-tile';
    }

    if (this.togglefull) {
      result += ' togglefull';
    }

    return result;
  }

  isToggled() {
    try {
      if (this.tyyppi) {
        const item = getItem('toggle-' + this.tyyppi);
        if (_.isObject(item)) {
          return (item as any).toggled;
        }
      }
      return true;
    }
    catch (err) {
      return true;
    }
  }

  mounted() {
    this.toggled = this.tyyppi
      ? this.isToggled()
      : this.expandedByDefault;
  }

  toggle(toggle: boolean | null = null) {
    if (this.collapsable) {
      if (!toggle) {
        this.toggled = !this.toggled;
      }
      else {
        this.toggled = toggle;
      }
      if (this.tyyppi) {
        setItem('toggle-' + this.tyyppi, {
          toggled: this.toggled,
        });
      }
    }
  }
}

</script>

<style scoped lang="scss">
@import '../../styles/_variables.scss';
@import '../../styles/_mixins.scss';

@include shadow-tile;

.topborder {
  border-top: 1px;
  border-top-style: solid;
  border-top-color: #eeeeee;
}

.bottomborder {
  border-bottom: 1px;
  border-bottom-style: solid;
  border-bottom-color: #eeeeee;
}

.ep-collapse {
  margin-top: 5px;

  .collapse-button {
    cursor: pointer;
    outline: none;

    label {
      cursor: pointer;
    }
  }

  .header-toggled {
    user-select: none;
  }

  &.togglefull {
    cursor: pointer;
  }
}

.shadow-tile {
  padding-left: 20px;
  padding-right: 20px;
  border-radius: 0.7rem;
}

::v-deep .osaamistasot {
  .row:nth-of-type(even) {
    background-color: $table-even-row-blue !important;
  }
  .row:nth-of-type(odd) {
    background-color: $table-odd-row-blue !important;
  }
}

::v-deep .table-responsive {
  tr:first-child td {
    background: $table-header-color !important;
  }
}

</style>
