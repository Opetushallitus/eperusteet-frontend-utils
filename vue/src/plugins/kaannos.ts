import Vue from 'vue';
import { createLogger } from '../utils/logger';
import { Kielet } from '../stores/kieli';
import * as _ from 'lodash';
import { unescapeStringHtml } from '@shared/utils/inputs';

const logger = createLogger('Kaannos');

declare module 'vue/types/vue' {
  interface Vue {
    $suodatin: (query: string) => <T extends object>(value: T) => string;
    $filterBy: (field: string, query: string) => <T extends object>(value: T) => boolean;
    $kaanna: <T extends object>(value?: T, emptyWhenNotFound?: boolean, squareBrackets?: boolean) => string;
    $kaannaOlioTaiTeksti: <T extends object>(value: T | string) => string;
  }
}

export class Kaannos {
  public install(vue: typeof Vue, options) {
    // Sisällön kääntäminen
    vue.prototype.$suodatin = (str: string) => Kielet.searchFn(str);

    vue.prototype.$filterBy = function(field: string, query: string) {
      return Kielet.filterBy(field, query);
    };

    vue.prototype.$kaanna = (value?: any, emptyWhenNotFound = false, squareBrackets = true) => {
      return this.handleUnescaping(Kielet.kaanna(value, emptyWhenNotFound, _.has(options, 'squareBrackets') ? options.squareBrackets : squareBrackets));
    };

    vue.prototype.$kaannaOlioTaiTeksti = (value: any, emptyWhenNotFound = false, squareBrackets = true) => {
      return this.handleUnescaping(Kielet.kaannaOlioTaiTeksti(
        value,
        _.has(options, 'emptyWhenNotFound') ? options.emptyWhenNotFound : emptyWhenNotFound,
        _.has(options, 'squareBrackets') ? options.squareBrackets : squareBrackets));
    };
  }

  // Ei tehdä unescapetusta, jos sisällössä oleva sisäinen linkki sisältää routenode-attribuutin, koska rikkoo tämän json-rakenteen.
  private handleUnescaping(kaannos) {
    return _.includes(kaannos, 'routenode') ? kaannos : unescapeStringHtml(kaannos);
  }
}

export default new Kaannos();
