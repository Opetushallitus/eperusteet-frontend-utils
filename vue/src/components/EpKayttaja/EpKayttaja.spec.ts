import { mount, createLocalVue, RouterLinkStub } from '@vue/test-utils';
import EpKayttaja from './EpKayttaja.vue';
import { Kieli } from '../../tyypit';
import { Kaannos } from '../../plugins/kaannos';
import VueI18n from 'vue-i18n';
import { delay } from '../../utils/delay';
import { Kielet } from '../../stores/kieli';
import Vue from 'vue';
import BootstrapVue from 'bootstrap-vue';
import EpCollapse from '../EpCollapse/EpCollapse.vue';

Vue.use(BootstrapVue);

describe('EpKayttaja component', () => {
  const localVue = createLocalVue();
  localVue.use(VueI18n);
  Kielet.install(localVue);
  localVue.use(new Kaannos());

  test('Renders', async () => {
    let currentRoute = {
      x: 42,
      params: {
        lang: 'fi',
      },
    };

    const wrapper = mount(EpKayttaja, {
      localVue,
      propsData: {
        tiedot: {
          kutsumanimi: 'etunimi',
          sukunimi: 'sukunimi',
        },
        sovellusOikeudet: [
          {
            eperusteSovellus:
              {
                sovellus: 'APP_EPERUSTEET',
                url: 'eperuste-url',
              },
            valittu: true,
          },
          {
            eperusteSovellus: {
              sovellus: 'APP_EPERUSTEET_AMOSAA',
              url: 'amosaa-url',
            },
            valittu: false,
          },
        ],
      },
      mocks: {
        $t: x => x,
        $router: {
          currentRoute,
          async push(route: any) {
            currentRoute = route;
          },
        },
      },
      stubs: {
        'router-link': RouterLinkStub,
      },
    });

    expect(wrapper.html()).toContain('etunimi sukunimi');
    expect(wrapper.find('.uikieli').html()).toContain('fi');
    wrapper.find(EpCollapse).trigger('click');
    await delay();
    await (wrapper.vm as any).valitseUiKieli(Kieli.sv);

    expect(currentRoute).toEqual({
      x: 42,
      params: {
        lang: 'sv',
      },
    });

    expect(wrapper.find('.valittu-sovellus').html()).toContain('APP_EPERUSTEET');
    wrapper.find('.valittu-sovellus').trigger('click');

    await delay();

    expect(wrapper.findAll(EpCollapse).at(1)
      .findAll('.sovellusoikeus').length).toBe(2);
    expect(wrapper.findAll(EpCollapse).at(1)
      .html()).toContain('amosaa-url');

    expect(wrapper.html()).toMatchSnapshot();
  });
});
