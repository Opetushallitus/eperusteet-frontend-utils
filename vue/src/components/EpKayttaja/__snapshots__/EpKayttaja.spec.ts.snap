// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpKayttaja component Renders 1`] = `
<div class="kayttaja">
  <li id="kayttaja-dropdown" class="nav-item b-nav-dropdown dropdown"><a id="kayttaja-dropdown__BV_toggle_" role="button" aria-haspopup="true" aria-expanded="false" href="#kayttaja-dropdown" target="_self" class="nav-link dropdown-toggle dropdown-toggle-no-caret">
      <div class="d-flex flex-row">
        <div class="kayttaja-valikko d-flex flex-column"><span class="kielivalitsin text-right">etunimi sukunimi</span>
          <!---->
        </div> <span aria-hidden="true" class="icon material-icons">expand_more</span>
      </div>
    </a>
    <ul tabindex="-1" aria-labelledby="kayttaja-dropdown__BV_toggle_" class="dropdown-menu dropdown-menu-right">
      <!---->
      <div>
        <!---->
        <div class="ep-collapse">
          <div role="button" tabindex="0" class="collapse-button d-flex align-items-center mb-2">
            <!---->
            <div class="align-self-start header">
              <div class="">
                <div>
                  <div class="pl-3 pt-2 text-nowrap kieli"><span aria-hidden="true" class="icon icon mr-3 material-icons">language</span> <span>kieli</span></div>
                  <div class="pl-3 uikieli"><span class="icon mr-3"></span> <small>sv</small></div>
                </div>
              </div>
            </div>
            <div class="ml-auto align-self-start" style="padding: 0.8rem 1rem;"><span aria-hidden="true" class="icon material-icons">expand_more</span></div>
          </div>
          <!---->
        </div>
        <!---->
      </div>
      <li role="presentation"><a role="menuitem" href="/henkilo-ui/omattiedot" target="_self" class="dropdown-item"><span aria-hidden="true" class="icon icon mr-3 material-icons">person</span> <span>kayttajan-asetukset</span></a></li>
      <li role="presentation">
        <hr role="separator" aria-orientation="horizontal" class="dropdown-divider">
      </li>
      <div>
        <!---->
        <div class="ep-collapse">
          <div role="button" tabindex="0" class="collapse-button d-flex align-items-center mb-2" aria-expanded="true">
            <!---->
            <div class="align-self-start header">
              <div class="header-toggled">
                <div>
                  <div class="pl-3 pt-2 text-nowrap kieli"><span aria-hidden="true" class="icon icon mr-3 material-icons">launch</span> <span>vaihda-sovellusta</span></div>
                  <div class="pl-3 valittu-sovellus pb-2"><span class="icon mr-3"></span> <small>APP_EPERUSTEET</small></div>
                </div>
              </div>
            </div>
            <div class="ml-auto align-self-start" style="padding: 0.8rem 1rem;"><span aria-hidden="true" class="icon material-icons">expand_less</span></div>
          </div>
          <div>
            <div class="collapse-tausta text-left">
              <li role="presentation" class="sovellusoikeus"><a role="menuitem" href="eperuste-url" target="_self" tabindex="-1" aria-disabled="true" class="dropdown-item disabled">
                  <div class="collapse-tausta-valinta-icon"><span aria-hidden="true" class="icon mr-3 valittu material-icons">done</span></div>
                  APP_EPERUSTEET
                </a></li>
              <li role="presentation" class="sovellusoikeus"><a role="menuitem" href="amosaa-url" target="_self" class="dropdown-item">
                  <div class="collapse-tausta-valinta-icon">
                    <!---->
                  </div>
                  APP_EPERUSTEET_AMOSAA
                </a></li>
              <li role="presentation"><a role="menuitem" href="/virkailijan-tyopoyta" target="_self" class="dropdown-item">
                  <div class="collapse-tausta-valinta-icon"></div>
                  virkailijan-tyopoyta
                </a></li>
            </div>
          </div>
        </div>
        <!---->
      </div>
      <li role="presentation">
        <hr role="separator" aria-orientation="horizontal" class="dropdown-divider">
      </li>
      <li role="presentation"><a role="menuitem" href="/service-provider-app/saml/logout" target="_self" class="dropdown-item"><span aria-hidden="true" class="icon mr-3 valittu material-icons">logout</span> <span>kirjaudu-ulos</span></a></li>
    </ul>
  </li>
</div>
`;
